### DISCUSSION TITLE
readFragment - Need help understanding why this example is returning null

### DISCUSSION THREAD

**Discussion by JohnAtFenestra** [urql-graphql/urql#1143] [Q&A] [answered]
We have a display containing a list of summary information about school visits. In some cases, the full information will be present in the cache; in other cases the user will not have accessed this information yet. The application has a full offline mode.

I am attempting to use the resolver to populate a flag field indicating whether or not the full information is in the cache. My test code is:

```javascript
  resolvers: {
    SampledGradeCard: {
      offlineMeta(parent, args, cache, info) {
        const data = cache.readFragment(
          gql`
            fragment _ on SampledGrade {
              visitId
            }
          `,
          parent.visitId
        )
        if (parent.visitId === '2410371+01/25/2021') console.log({ data })

        return data === null ? 'not-in-cache' : 'in-cache'
      }
    }
  },
```

Keys are set as follows (abridged):

```javascript
  keys: {
    SampledGrade: data => data.visitId,
  }
```

A small extract of the cache:

```javascript
{
  "SampledGrade:2410371+01/25/2021.__typename": "\"SampledGrade\"",
  "SampledGrade:2410371+01/25/2021.visitId": "\"2410371+01/25/2021\"",
```

The debug test shows that the visit id I've hardcoded is in the incoming values, but the `readFragment` command doesn't return the value - it returns `null`. Can you help me understand what I'm doing wrong with the `readFragment` command?

**Comment by kitten** [comment, marked as answer]
`cache.readFragment`’s second argument here needs to be a key or a keyable entity, so `{ __typename: "SampledGrade", visitId: "..." } ` if that's your full type or `cache.keyOfEntity({ __typename: "SampledGrade", visitId: "..." })`.

As you can see here by `keyOfEntity` is what you may be confused about. A key is not an ID, because we still preserve the type in our key. Essentially that's opinionated but unlike Relay we assert that no two types can share the same entity as they're different, so all keys contain the type name.

Given that and the method the `readFragment` method only accepts a full key or a partial entity from which a key can be generated.

By the way, given this there's a shorter way to do this. You may also use `cache.resolve` to assert whether an entity exists using `!!cache.resolve({ __typename: "SampleGrade", visitId }, "visitId")` which works because it's asking for this specific property existing on the cache.

**Comment by JohnAtFenestra** [comment]
Thank you @kitten. I was confused by the example in the docs:

```javascript
import gql from 'graphql-tag';
const data = cache.readFragment(
  gql`
    fragment _ on Todo {
      id
      text
    }
  `,
  '1'
);
```

Your explanation helps a lot!

Do you know if `keyOfEntity` is documented anywhere except the source? I don't see it in the Graphcache documentation.

**Reply by kitten** [reply]
It's in the API docs: https://formidable.com/open-source/urql/docs/api/graphcache/#cache

The methods are also more thoroughly described there

**Comment by JohnAtFenestra** [comment]
Thanks!
